Refactor CMake code for dictionary regeneration

Teemperor opened the pull request:
Just to test in jenkins

Teemperor commented:
Still needs a lot of formatting, will do when I have time.

martinmine commented:
@phsft-bot build and stop being lazy

Teemperor commented:
@phsft-bot build

amadio commented on cmake/modules/RootNewMacros.cmake:
```diff
@@ -827,13 +827,70 @@ function(ROOT_INSTALL_HEADERS)
+    set(ARG_HEADERS "*.h")
+  endif()
+  if (NOT ARG_SOURCES AND NOT ARG_HEADER_ONLY)
+    set(ARG_SOURCES "*.cxx")
```
Should this be "src/*.cxx"?

Teemperor replied:
Yeah, but I didn't want to change ROOT_STANDARD_LIBRARY_PACKAGE behavior (right now it behaves exactly the same way as before as long as you don't use the new arguments). Maybe we should indeed use a new function. Let's discuss this on monday.

Teemperor commented:
@phsft-bot build

jenkins crashed during checkout...

Teemperor commented:
@phsft-bot build

Teemperor commented:
@phsft-bot build

vgvassilev commented:
Can you rename the commit message. Looks good overall, let me know if I should test it on massively parallel machine.

peremato commented on cmake/modules/RootNewMacros.cmake:
```diff
@@ -820,13 +820,77 @@ function(ROOT_INSTALL_HEADERS)
+#---ROOT_STANDARD_LIBRARY_PACKAGE(libname
+#                                 [NO_INSTALL_HEADERS]         : don't install headers for this package
+#                                 [STAGE1]                     : use rootcling_stage1 for generating
+#                                 HEADER header1 header2       : if not specified, globbing for *.h is used)
```
The parameter is HEADERS and not HEADER

peremato commented on cmake/modules/RootNewMacros.cmake:
```diff
@@ -820,13 +820,77 @@ function(ROOT_INSTALL_HEADERS)
-  ROOT_GENERATE_DICTIONARY(G__${libname} *.h MODULE ${libname} LINKDEF LinkDef.h OPTIONS ${ARG_DICTIONARY_OPTIONS})
-  ROOT_LINKER_LIBRARY(${libname} *.cxx G__${libname}.cxx DEPENDENCIES ${ARG_DEPENDENCIES})
-  ROOT_INSTALL_HEADERS()
+  set(options OPTIONAL NO_INSTALL_HEADERS STAGE1 NO_HEADERS NO_SOURCES OBJECT_LIBRARY)
```
Is OPTIONAL used anywhere?

peremato commented on geom/geocad/CMakeLists.txt:
```diff
@@ -8,8 +8,5 @@ include_directories(${OCC_INCLUDE_DIR})
-ROOT_LINKER_LIBRARY(GeoCad *.cxx G__GeoCad.cxx LIBRARIES Core ${OCC_LIBRARIES} DEPENDENCIES Geom)
-
-ROOT_INSTALL_HEADERS()
+ROOT_STANDARD_LIBRARY_PACKAGE(GeoCad HEADERS *.h DEPENDENCIES Geom LIBRARIES ${OCC_LIBRARIES})
```
probably HEADERS *.h is not needed

peremato commented on graf2d/graf/CMakeLists.txt:
```diff
@@ -3,11 +3,7 @@
 include_directories(${FREETYPE_INCLUDE_DIRS})
 
-ROOT_LINKER_LIBRARY(Graf *.cxx G__Graf.cxx LIBRARIES ${FREETYPE_LIBRARIES} ${ZLIB_LIBRARIES} mathtext DEPENDENCIES Hist Matrix MathCore RIO BUILTINS FREETYPE)
+ROOT_STANDARD_LIBRARY_PACKAGE(Graf HEADERS *.h DICTIONARY_OPTIONS "-writeEmptyRootPCM" LIBRARIES ${FREETYPE_LIBRARIES} ${ZLIB_LIBRARIES} mathtext DEPENDENCIES Hist Matrix MathCore RIO BUILTINS FREETYPE)
```
idem HEADERS *.h is probably not needed (default)

peremato commented on graf3d/gl/CMakeLists.txt:
```diff
@@ -37,8 +37,4 @@ endif()
-ROOT_LINKER_LIBRARY(RGL ${sources} G__RGL.cxx  LIBRARIES ${gllibs} ${GL2PS_LIBRARIES} GLEW ${FTGL_LIBRARIES} DEPENDENCIES Hist Gui Ged)
-
-ROOT_INSTALL_HEADERS(${installoptions})
+ROOT_STANDARD_LIBRARY_PACKAGE(RGL HEADERS ${headers} SOURCES ${sources} DICTIONARY_OPTIONS "-writeEmptyRootPCM" LIBRARIES ${gllibs} ${GL2PS_LIBRARIES} GLEW ${FTGL_LIBRARIES} DEPENDENCIES Hist Gui Ged INSTALL_OPTIONS ${installoptions})
```
General comment. Some of these lines are too long. In some cases you have formatted nicely like:
```
ROOT_STANDARD_LIBRARY_PACKAGE(RGL HEADERS ${headers} 
                                  SOURCES ${sources} 
                                  DICTIONARY_OPTIONS "-writeEmptyRootPCM" 
                                  LIBRARIES ${gllibs} ${GL2PS_LIBRARIES} GLEW ${FTGL_LIBRARIES} 
                                  DEPENDENCIES Hist Gui Ged 
                                  INSTALL_OPTIONS ${installoptions})
```

amadio commented:
@peremato We will test everything on several places before merging. The previous changes were tested by me on my laptop, on a server with 36 cores, and on a KNL machine. I used 6, 70, and 250 threads, respectively, and had no failures. The same testing will happen for this PR to ensure that nothing breaks. Then we merge and check that no failures show up on the nightlies and incrementals.

peremato commented:
@amadio what I wanted to say is that it is not sufficient to run all the tests, which are run in the build directory and I assume will be done correctly, we need to check that the installation (make install) are complete and no header has been forgotten in the process. The new generic macro deals with the header installation and this is why we need to check carefully.

vgvassilev commented:
Probably one easy way to assert that is to diff the modulemap files before and after this patch. They work with the headers to be installed, too.

amadio commented:
We could also create a diff of a list of all files installed.

vgvassilev commented:
ping...

peremato commented:
Ping what? I reviewed 8 days ago.

vgvassilev commented:
@peremato, it was a ping for @Teemperor.

Teemperor commented:
Updated, fixed a typo in one of the windows-only CMake files, fixed a few more differences between what headers are installed and formatted everything. Can someone give it a test install to double-check the headers?

amadio commented:
I created a tarball with the installed files from the master branch and this branch and I see no difference when I apply a diff on the list of installed files. @vgvassilev Feel free to merge.
